elastic/apm-server
# apmpackage: define DLM retention periods in apm package

## Motivation/summary

Define DLM retention periods corresponding to the default ILM policy's `policy.delete.min_age` to support DLM when configured.

## Checklist



~- [ ] Documentation has been updated~

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## How to test these changes


1. Run `tilt up`
1. Open Kibana, go to Index Management > Index Template and choose one of the template apm server manages, for example: `metrics-apm.service_summary.1m`
1. Check the lifecycle is pointing to the ILM policy
![image](https://github.com/elastic/apm-server/assets/16660866/94780602-c006-49e9-ac6c-b2ab8b59bdfa)
1. Run `tilt down`, add `xpack.fleet.internal.disableILMPolicies: true` to the _**config**_ section in the `testing/infra/k8s/base/stack/kibana.yaml` file. (This mimics the serverless config https://github.com/elastic/kibana/pull/154460/files#diff-ebeb9e9dc4c3ee5e81f743f0eda3c06f1fc161a11127b374d21f0da5e6368fd9R2)
1. Run `tilt up` again, navigate to Index Management, and observe the lifecycle is now using `data_rentention` rather than the ILM policy.
![image](https://github.com/elastic/apm-server/assets/16660866/08875b1f-0871-47a6-8798-2a245975af34)


1. Now, modify the retention period in `apmpackage/apm/data_stream/service_summary_interval_metrics/lifecycle.1m.yml` to `1m`, and when the `apmpackage` is reinstalled, generate some load so that the datastream is created.
1. Go to Dev Tool, and confirm the `data_retention` for the datastream is correctly set by running the following:
```
GET _data_stream/metrics-apm.service_summary.1m-default/_lifecycle
```
Should see
```
{
  "data_streams": [
    {
      "name": "metrics-apm.service_summary.1m-default",
      "lifecycle": {
        "enabled": true,
        "data_retention": "1m"
      }
    }
  ]
}
```

8. See the documents are all deleted after a while (it's not precisely 1 minute to get deleted, so you might need to wait a bit longer than the data_retention period)

## Related issues
- https://github.com/elastic/kibana/pull/161905
- https://github.com/elastic/package-spec/pull/556
